File: intrinsic/skills/internal/skill_service_config_utils.py
# Copyright 2023 Intrinsic Innovation LLC

"""Utils for working with the SkillServiceConfig proto."""

from absl import logging
from google.protobuf import descriptor_pb2
from intrinsic.skills.internal import proto_utils
from intrinsic.skills.proto import skill_manifest_pb2
from intrinsic.skills.proto import skill_service_config_pb2


def get_skill_service_config(
    filename: str,
) -> skill_service_config_pb2.SkillServiceConfig:
  """Returns the SkillServiceConfig at `skill_service_config_filename`.

  Reads the proto binary file at `skill_service_config_filename` and returns
  the contents. This file must contain a proto binary
  intrinsic_proto.skills.SkillServiceConfig message.

  Args:
    filename: the filename of the binary proto file

  Returns:
    A skill_service_config_pb2.SkillServiceConfig message
  """
  logging.info(
      "Reading skill configuration proto from: %s",
      filename,
  )
  service_config = skill_service_config_pb2.SkillServiceConfig()
  logging.info("\nImporting service_config from file.")
  with open(filename, "rb") as f:
    service_config.ParseFromString(f.read())
  logging.info("\nUsing skill configuration proto:\n%s", service_config)
  return service_config


def get_skill_service_config_from_manifest(
    manifest_pbbin_filename: str,
    proto_descriptor_pbbin_filename: str,
    version: str,
) -> skill_service_config_pb2.SkillServiceConfig:
  """Returns the SkillServiceConfig for the skill with the given manifest."""
  service_config = skill_service_config_pb2.SkillServiceConfig()

  logging.info("Loading manifest from: %s", manifest_pbbin_filename)
  with open(manifest_pbbin_filename, "rb") as f:
    manifest = skill_manifest_pb2.Manifest.FromString(f.read())

  service_config.skill_name = manifest.id.name
  if manifest.options.HasField("cancellation_ready_timeout"):
    service_config.execution_service_options.cancellation_ready_timeout.CopyFrom(
        manifest.options.cancellation_ready_timeout
    )
  if manifest.options.HasField("execution_timeout"):
    service_config.execution_service_options.execution_timeout.CopyFrom(
        manifest.options.execution_timeout
    )

  service_config.status_info.extend(manifest.status_info)

  logging.info(
      "Reading proto descriptor proto from: %s",
      proto_descriptor_pbbin_filename,
  )
  with open(proto_descriptor_pbbin_filename, "rb") as f:
    file_descriptor_set = descriptor_pb2.FileDescriptorSet.FromString(f.read())

  service_config.skill_description.CopyFrom(
      proto_utils.proto_from_skill_manifest(
          manifest, file_descriptor_set, version=version
      )
  )

  return service_config

File: intrinsic/skills/internal/skill_service_config_utils.cc
// Copyright 2023 Intrinsic Innovation LLC

#include "intrinsic/skills/internal/skill_service_config_utils.h"

#include <optional>

#include "absl/log/log.h"
#include "absl/status/statusor.h"
#include "absl/strings/string_view.h"
#include "google/protobuf/descriptor.pb.h"
#include "intrinsic/assets/proto/id.pb.h"
#include "intrinsic/assets/proto/status_spec.pb.h"
#include "intrinsic/icon/release/file_helpers.h"
#include "intrinsic/skills/internal/skill_proto_utils.h"
#include "intrinsic/skills/proto/skill_manifest.pb.h"
#include "intrinsic/skills/proto/skill_service_config.pb.h"
#include "intrinsic/util/status/status_macros.h"

namespace intrinsic::skills {

absl::StatusOr<intrinsic_proto::skills::SkillServiceConfig>
GetSkillServiceConfig(absl::string_view skill_service_config_filename) {
  intrinsic_proto::skills::SkillServiceConfig service_config;
  if (!skill_service_config_filename.empty()) {
    LOG(INFO) << "Reading skill configuration proto from "
              << skill_service_config_filename;
    INTR_ASSIGN_OR_RETURN(
        service_config,
        intrinsic::GetBinaryProto<intrinsic_proto::skills::SkillServiceConfig>(
            skill_service_config_filename));
    LOG(INFO) << "\nUsing skill configuration proto:\n" << service_config;
  }
  return service_config;
}

absl::StatusOr<intrinsic_proto::skills::SkillServiceConfig>
GetSkillServiceConfigFromManifest(
    absl::string_view manifest_pbbin_filename,
    absl::string_view file_descriptor_set_pbbin_filename,
    std::optional<absl::string_view> version) {
  intrinsic_proto::skills::SkillServiceConfig service_config;

  LOG(INFO) << "Loading manifest from " << manifest_pbbin_filename;
  INTR_ASSIGN_OR_RETURN(
      auto manifest,
      intrinsic::GetBinaryProto<intrinsic_proto::skills::SkillManifest>(
          manifest_pbbin_filename));

  LOG(INFO) << "Loading FileDescriptorSet from "
            << file_descriptor_set_pbbin_filename;
  INTR_ASSIGN_OR_RETURN(
      auto file_descriptor_set,
      intrinsic::GetBinaryProto<google::protobuf::FileDescriptorSet>(
          file_descriptor_set_pbbin_filename));

  return GetSkillServiceConfigFromManifest(manifest, file_descriptor_set,
                                           version);
}

absl::StatusOr<intrinsic_proto::skills::SkillServiceConfig>
GetSkillServiceConfigFromManifest(
    const intrinsic_proto::skills::SkillManifest& manifest,
    const google::protobuf::FileDescriptorSet& file_descriptor_set,
    std::optional<absl::string_view> version) {
  intrinsic_proto::skills::SkillServiceConfig service_config;

  service_config.set_skill_name(manifest.id().name());

  if (manifest.options().has_cancellation_ready_timeout()) {
    *service_config.mutable_execution_service_options()
         ->mutable_cancellation_ready_timeout() =
        manifest.options().cancellation_ready_timeout();
  }
  if (manifest.options().has_execution_timeout()) {
    *service_config.mutable_execution_service_options()
         ->mutable_execution_timeout() = manifest.options().execution_timeout();
  }

  *service_config.mutable_status_info() = manifest.status_info();

  INTR_ASSIGN_OR_RETURN(
      *service_config.mutable_skill_description(),
      BuildSkillProto(manifest, file_descriptor_set, version));

  return service_config;
}

absl::StatusOr<intrinsic_proto::skills::SkillServiceConfig>
GetSkillServiceConfigFromManifest(
    const intrinsic_proto::skills::SkillManifest& manifest,
    const google::protobuf::FileDescriptorSet& parameter_descriptor_set,
    const google::protobuf::FileDescriptorSet& return_type_descriptor_set,
    std::optional<absl::string_view> version) {
  intrinsic_proto::skills::SkillServiceConfig service_config;

  service_config.set_skill_name(manifest.id().name());

  if (manifest.options().has_cancellation_ready_timeout()) {
    *service_config.mutable_execution_service_options()
         ->mutable_cancellation_ready_timeout() =
        manifest.options().cancellation_ready_timeout();
  }
  if (manifest.options().has_execution_timeout()) {
    *service_config.mutable_execution_service_options()
         ->mutable_execution_timeout() = manifest.options().execution_timeout();
  }

  *service_config.mutable_status_info() = manifest.status_info();

  INTR_ASSIGN_OR_RETURN(*service_config.mutable_skill_description(),
                        BuildSkillProto(manifest, parameter_descriptor_set,
                                        return_type_descriptor_set, version));

  return service_config;
}

}  // namespace intrinsic::skills
